### PR TITLE
Numpy version of point_cloud2.read_points_list and point_cloud2.read_points

### PR DESCRIPTION
**Pull request by dheera**
In the spirit of ROS2 which delivers messages with ndarray / array instead of Python lists, this avoids slow Python iteration over points. In this new implementation there are no python loops over actual PointCloud2 points, everything is vectorized.

It works by simply "typecasting" the PointCloud2 data into a struct array using numpy views and recarrays.

### CODE REVIEW

**Comment by dheera** [comment]
I couldn't get the DCO working. Not sure what that is. I tried to follow the directions and I get
```
:) git rebase HEAD~3 --signoff
error: cannot rebase: Your index contains uncommitted changes.
error: Please commit or stash them.
```

I tried to follow the directions in the referenced link (https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally) and I get this
```
:) git fetch origin pull/155/head:master
fatal: couldn't find remote ref pull/155/head
```

**Comment by dheera** [comment]
I tried this and got a merge commit this time.
```
:) git rebase HEAD~4 --signoff
Current branch master is up to date, rebase forced.
First, rewinding head to replay your work on top of it...
Applying: numpy-ified point_cloud2.py
Applying: map all numpy.record to python lists before testing
Applying: add author line
Applying: fix style issues
(base) ~/Dropbox/code/ros/common_interfaces/sensor_msgs_py :) git push
To github.com:/dheera/common_interfaces
 ! [rejected]        master -> master (non-fast-forward)
error: failed to push some refs to 'git@github.com:/dheera/common_interfaces'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

**Comment by dheera** [comment]
I then tried
```
 :) git commit --amend --signoff
[master 1a11e58] Merge branch 'master' of github.com:/dheera/common_interfaces
 Date: Sat Jul 17 23:47:08 2021 +0000
```
and then it punished me with 2 more merge commits. This is quite unforgiving!

**Comment by flynneva** [comment]
> error: cannot rebase: Your index contains uncommitted changes.
error: Please commit or stash them.

This looks like you have uncommitted local changes. run 'git status' before rebasing and make sure you don't have any uncommitted changes.

**Comment by flynneva** [comment]
It looks like you've duplicated your commits and there is only 4 (maybe 3?) "real" commits for this PR.

I'd recommend resetting back to there and force push to reset your branch without the redundant commits. This should remove your extra merge commits too.

**Comment by dheera** [comment]
@flynneva Is it possible to reset back without re-creating the branch? Every time I've git reset before it created more commits.

**Comment by flynneva** [comment]
Yeah, three steps but you can do it. "git reset HEAD~8" to reset back 8 commits (or however many you want to go back), "git restore ." to get rid of any changes that might still be in your local repo. "git push -f" to force push your new commit history.

Obviously this should only be done in unique situations since you are essentially erasing some git history that is out in the world. In this situation though it makes sense I think.

File: sensor_msgs_py/sensor_msgs_py/point_cloud2.py [hunk 1]
```diff
@@ -68,80 +81,72 @@ def read_points(cloud, field_names=None, skip_nans=False, uvs=[]):
...
+    points = np.frombuffer(cloud.data, dtype=np_struct).view(dtype=np.recarray)
 
-    Point = namedtuple('Point', field_names)
+    if skip_nans:
+        nan_indexes = None
+        for field_name in all_field_names:
+            if nan_indexes is None:
+                nan_indexes = np.isnan(points[field_name])
+            else:
+                nan_indexes = nan_indexes | np.isnan(points[field_name])
```
Comment (hidmic) [inline sensor_msgs_py/sensor_msgs_py/point_cloud2.py:134]: @dheera nit^2: 
```suggestion
        nan_indexes = False
        for field_name in all_field_names:
            nan_indexes |= np.isnan(points[field_name])
```

Comment (hidmic) [reply sensor_msgs_py/sensor_msgs_py/point_cloud2.py:134]: Also, if `field_names` is not `None`, we can probably filter columns out before doing this.

**Review by hidmic — COMMENTED** [review]
@dheera DCO still needs to be fixed up. Also, would you mind adding a test to check the record-like structure of points? E.g. grab a point, check expected attributes are there.

**Comment by aprotyas** [comment]
About the issues with DCO, I would suggest you just squash all your commits with `git rebase -i HEAD~[X]`, where `X` is the number of commits you want to squash. Just make sure you sign the squashed commit.

**Comment by flynneva** [comment]
@dheera do you care if I make another PR for this change to take this off your plate?

**Comment by clalancette** [comment]
Closing in favor of #175.

**Pull request closed without merging**